### PR TITLE
[WIP] CODENVY-627 Restrict using reserved names in user names

### PR DESCRIPTION
**Pull request by mkuznyetsov**

### CODE REVIEW

File: core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserCreator.java [hunk 1]
```diff
@@ -145,4 +145,19 @@ public User createTemporary() throws IOException {
...
+
+    private boolean createNonReservedUser(String id, String username, String email, String password) throws ServerException{
+        try {
+            userManager.create(new User().withId(id)
+                                         .withName(username)
+                                         .withEmail(email)
+                                         .withPassword(password), false);
+            return true;
+        } catch (ServerException e) {
+            e.printStackTrace();
```
Comment (sleshchenko) [inline core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserCreator.java]: Are you sure that we need this?

File: core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserCreator.java [hunk 2]
```diff
@@ -81,19 +82,18 @@ public User createUser(String email, String userName, String firstName, String l
...
 
             try {
-                final User user = new User().withId(id)
-                                            .withName(userName)
-                                            .withEmail(email)
-                                            .withPassword(password);
-                userDao.create(user);
+
+                while (!createNonReservedUser(id, userName, email, password)) {
+                    NameGenerator.generate(userName, 4);
```
Comment (benoitf) [inline core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserCreator.java]: we shouldn't get value of the call to generate ?

File: core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserCreator.java [hunk 3]
```diff
@@ -81,19 +82,18 @@ public User createUser(String email, String userName, String firstName, String l
             String password = UUID.randomUUID().toString().replace("-", "").substring(0, 12);
 
             try {
-                final User user = new User().withId(id)
-                                            .withName(userName)
-                                            .withEmail(email)
-                                            .withPassword(password);
-                userDao.create(user);
+
+                while (!createNonReservedUser(id, userName, email, password)) {
```
Comment (benoitf) [inline core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserCreator.java]: a test is required to validate when it fails and if retries work as expected

File: core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserCreator.java (hunk 1 above)
Comment (mkuznyetsov) [inline core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserCreator.java]: will fix

File: core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserCreator.java (hunk 2 above)
Comment (mkuznyetsov) [inline core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserCreator.java]: will fix

File: core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserCreator.java (hunk 3 above)
Comment (mkuznyetsov) [inline core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserCreator.java]: ok

File: core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserValidator.java [hunk 4]
```diff
@@ -36,14 +40,17 @@
...
     @Inject
     public OrgServiceUserValidator(UserDao userDao,
                                    UserNameValidator userNameValidator,
-                                   @Named("user.self.creation.allowed") boolean userSelfCreationAllowed) {
+                                   @Named("user.self.creation.allowed") boolean userSelfCreationAllowed,
+                                   @Named("user.reserved_names") String[] reservedNames) {
         this.userDao = userDao;
         this.userNameValidator = userNameValidator;
         this.userSelfCreationAllowed = userSelfCreationAllowed;
+        this.reservedNames = Sets.newHashSet(Arrays.asList(reservedNames));
```
Comment (benoitf) [inline core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserValidator.java]: with guava you could give the array directly
http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/collect/Sets.html#newHashSet(E...)

Comment (mkuznyetsov) [inline core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserValidator.java]: ok

File: core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserCreator.java [hunk 5]
```diff
@@ -145,4 +147,18 @@ public User createTemporary() throws IOException {
             throw new IOException(e.getLocalizedMessage(), e);
         }
     }
+
+    private boolean createNonReservedUser(String id, String username, String email, String password) throws ServerException{
```
Comment (benoitf) [inline core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserCreator.java]: missing space before last {

File: core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserValidator.java [hunk 6]
```diff
@@ -64,6 +71,10 @@ public void ensureUserCreationAllowed(String email, String userName) throws Conf
             throw new BadRequestException("User name must contain letters and digits only");
         }
 
+        if (reservedNames.contains(userName)) {
+            throw new ConflictException("User name is reserved. Please, choose another one");
```
Comment (benoitf) [inline core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserValidator.java]: include username -->
throw new ConflictException(String.format("User %s is reserved, Please, choose another one", userName))

File: core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserCreator.java (hunk 5 above)
Comment (mkuznyetsov) [inline core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserCreator.java]: fixed

File: core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserValidator.java (hunk 6 above)
Comment (mkuznyetsov) [inline core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserValidator.java]: ok

File: core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserCreator.java (hunk 3 above)
Comment (benoitf) [inline core/codenvy-hosted-sso-server-organization/src/main/java/com/codenvy/auth/sso/server/OrgServiceUserCreator.java]: on a side note, maybe we should limit the number of "retries"
if for example the server is throwing exception at each call, we will enter into an infinite loop

**Comment by benoitf** [comment]
ok for me

File: core/codenvy-hosted-sso-server-organization/src/test/java/com/codenvy/auth/sso/server/OrgServiceUserCreatorTest.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    UserProfileDao profileDao;
+
+    @Mock
+    PreferenceDao preferenceDao;
+
+    @Test
+    public void shouldCreateUser() throws Exception {
+        doThrow(NotFoundException.class).when(manager).getByAlias(anyObject());
+
+        new OrgServiceUserCreator(manager, profileDao, preferenceDao, true).createUser("user@codenvy.com", "test", "John", "Doe");
```
Comment (skabashnyuk) [inline core/codenvy-hosted-sso-server-organization/src/test/java/com/codenvy/auth/sso/server/OrgServiceUserCreatorTest.java]: new OrgServiceUserCreator(manager, profileDao, preferenceDao, true) 
setup out of test

Comment (mkuznyetsov) [inline core/codenvy-hosted-sso-server-organization/src/test/java/com/codenvy/auth/sso/server/OrgServiceUserCreatorTest.java]: ok

**Comment by skabashnyuk** [comment]
ок, wait for feature approval before merge

**Pull request closed without merging**